src/gsi/gsi_rfv3io_mod.f90

```diff
@@ -2372,193 +2338,98 @@ subroutine gsi_fv3ncdf_write(filename,varname,var,mype_io,add_saved)
 !
 !$$$ end documentation block
 
-    use mpimod, only: mpi_rtype,mpi_comm_world,ierror,npe,mype
-    use gridmod, only: lat2,lon2,nlon,nlat,lat1,lon1,nsig
-    use gridmod, only: ijn,displs_g,itotsub,iglobal
+    use mpimod, only: mpi_rtype,mpi_comm_world,mype,mpi_info_null
     use mod_fv3_lola, only: fv3_ll_to_h
     use mod_fv3_lola, only: fv3_h_to_ll
-    use general_commvars_mod, only: ltosi,ltosj
     use netcdf, only: nf90_open,nf90_close
     use netcdf, only: nf90_write,nf90_inq_varid
     use netcdf, only: nf90_put_var,nf90_get_var
+    use gsi_bundlemod, only: gsi_bundle
+    use general_sub2grid_mod, only: sub2grid_info,general_sub2grid
     implicit none
+    type(sub2grid_info), intent(in):: grd_ionouv 
+    type(gsi_bundle),intent(inout) :: cstate_nouv
 
-    real(r_kind)   ,intent(in   ) :: var(lat2,lon2,nsig)
-    integer(i_kind),intent(in   ) :: mype_io
     logical        ,intent(in   ) :: add_saved
-    character(*)   ,intent(in   ) :: varname,filename
+    character(len=:), allocatable, intent(in) :: filenamein
+    type (type_fv3regfilenameg),intent (in) :: fv3filenamegin
+    real(r_kind),dimension(1,grd_ionouv%nlat,grd_ionouv%nlon,grd_ionouv%kbegin_loc:grd_ionouv%kend_alloc):: hwork
+    character(len=max_varname_length) :: filenamein2 
+    character(len=max_varname_length) :: varname,vgsiname
 
+    integer(i_kind) nlatcase,nloncase,nxcase,nycase,countloc(3),startloc(3)
+    integer(i_kind) kbgn,kend
+    integer(i_kind) inative,ilev,ilevtot
     integer(i_kind) :: VarId,gfile_loc
-    integer(i_kind) i,j,mm1,k,kr,ns,n,m
-    real(r_kind),allocatable,dimension(:):: work
-    real(r_kind),allocatable,dimension(:,:,:):: work_sub,work_a
-    real(r_kind),allocatable,dimension(:,:,:):: work_b
+    integer(i_kind) mm1,nzp1
+    real(r_kind),allocatable,dimension(:,:):: work_a
+    real(r_kind),allocatable,dimension(:,:):: work_b
     real(r_kind),allocatable,dimension(:,:):: workb2,worka2
 
 
     mm1=mype+1
+    ! Convert from subdomain to full horizontal field distributed among
+    ! processors
+    call general_sub2grid(grd_ionouv,cstate_nouv%values,hwork)
+    nloncase=grd_ionouv%nlon
+    nlatcase=grd_ionouv%nlat
+    nxcase=nx
+    nycase=ny
+    kbgn=grd_ionouv%kbegin_loc
+    kend=grd_ionouv%kend_loc
+    allocate( work_a(nlatcase,nloncase))
+    allocate( work_b(nlon_regional,nlat_regional))
+    allocate( workb2(nlon_regional,nlat_regional))
+    allocate( worka2(nlatcase,nloncase))
+    call check( nf90_open(filenamein,nf90_write,gfile_loc,comm=mpi_comm_world,info=MPI_INFO_NULL) )
+    do ilevtot=kbgn,kend
+       vgsiname=grd_ionouv%names(1,ilevtot)
+       call getfv3lamfilevname(vgsiname,fv3filenamegin,filenamein2,varname)
+       if(trim(filenamein).ne.trim(filenamein2)) then
+        write(6,*)'filenamein and filenamein2 are not the same as expected, stop'
+        call flush(6)
+        call stop2(333)
+       endif
```

Comment:
Please replace .ne. with /= and please apply proper indentation to lines 2388 through 2392.